### DISCUSSION TITLE
field "TypeMeta" not allowed in closed struct

### DISCUSSION THREAD

**Discussion by errordeveloper** [cuelang/cue#439] [General]
I have am using CUE in a Kubernetes operator, I am trying to pass my API type to CUE, and seeing this error:
```
cue: marshal error: template.items.0.metadata.name: field "TypeMeta" not allowed in closed struct
```

My template looks like this:

```
package basic

import "<path>/operator/api/v1alpha1"

MyTemplate :: {
	kind: "List"
	apiVersion: "v1"
	items: [{
		metadata: {
			name:      "\(resource.metadata.name)"
			namespace: "\(resource.metadata.namespace)"
		}
        }]
}

resource: v1alpha1.MyType

template: MyTemplate
```

The generated CUE types look like this:
```
MyType :: {
	metav1.TypeMeta
	metadata?: metav1.ObjectMeta    @go(ObjectMeta)
	spec?:     MyTypeSpec   @go(Spec)
	status?:   MyTypeStatus @go(Status)
}
```

I tried changing the type definition by removing `metav1.TypeMeta`, and the error I got was this:
```
cue: marshal error: template.items.3.spec.networkRef.name: field "name" is optional
```

I figured I'd try adding default values in my CUE tempate:
```
package basic

import "<path>/operator/api/v1alpha1"

MyTemplate :: {
	kind: "List"
	apiVersion: "v1"
	items: [{
		metadata: {
			name:      "\(resource.metadata.name)" | ""
			namespace: "\(resource.metadata.namespace)" | ""
		}
        }]
}

resource: v1alpha1.MyType

template: MyTemplate
```

That worked. I then added `metav1.TypeMeta` back and all works now.

I don't really understand why is this happening, could someone explain? What can I do to avoid having default values in the template?

I am pretty sure that `TypeMeta` error is a misleading actually, it only seems happens when an optional value has no default.

Here is a more elegant work-around:
```
package basic

import "<path>/operator/api/v1alpha1"

MyTemplate :: {
	kind: "List"
	apiVersion: "v1"
	items: [{
		metadata: {
			name:      "\(resource.metadata.name)" | "\(defaults.metadata.name)" 
			namespace: "\(resource.metadata.namespace)" | "\(defaults.metadata.namespace)" 
		}
        }]
}

resource: v1alpha1.MyType

template: MyTemplate

defaults: {}
```
I have a separate `Fill` call that sets `defaults`, it still uses `v1alpha1.MyType`, but somehow not having the type declared in the template makes it work. Albeit, as soon as one of the fields is not set in `defaults`, I get my `TypeMeta` error back again.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/439.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
